pkg/util/util.go

```diff
@@ -82,6 +82,10 @@ type DownloadParams struct {
 }
 
 // ConvertLabelsToSelector converts the given labels to selector
+// To pass operands such as !=, append a ! prefix to the value.
+// For E.g. map[string]string{"app.kubernetes.io/managed-by": "!odo"}
+// Using != operators also means that resource will be filtered even if it doesn't have the key.
+// So a resource not labelled with key "app.kubernetes.io/managed-by" will also be returned.
 func ConvertLabelsToSelector(labels map[string]string) string {
```

Comment:
why extend the labels here to be a list of structs here like 
```
struct {
Key string
Value string
Operand Operand
}

type Operand string

var Equal Operand = "="
var NotEqual Operand = "!="
```

might be more future proof?

Reply:
Can you elaborate?

Reply:
instead of this label map
```
labels map[string]string
```


```
struct {
Key string
Value string
Operand Operand
}

type Operand string

var Equal Operand = "="
var NotEqual Operand = "!="

```

So the function would look like this `ConvertLabelsToSelector(labels []Label)`.

So I can do something like this 
```
labels := []Label{
   {
   Key: "app.kubernetes.io/managed-by",
   Operand: NotEqual,
    Value: "odo"
     }
}

ConvertLabelsToSelector(labels)
```

The conversion of `[]Label` to `string` would also become simplified

Reply:
This would allow for room for extension and also less mistakes

Reply:
It seems to me as if this will require a significant amount of effort and it seems a little out of scope for this PR. Is it alright if I take this in another PR?

Also, can you give me an example of possible mistakes that could happen here?

Reply:
e.g. someone did `!!odo` the logic of `=` and `!=` is dependent on `!` and also it doesn't come as natural to the reader of the code. they have to go to the function and read its comment ( which is fine ) but says that there is room for improvement.
Also the `Label` approach allows for simpler extending in future.

Reply:
But about not being in scope, I am not sure about that as the logic of `!` was introduced in this PR 🙂 . So the refactor for the code would probably be part of this PR too?

Reply:
@valaparthvi I'm not sure why someone would try to set `"app.kubernetes.io/managed-by": "!odo"` as a label to the Kubernetes resources created by odo. At least in the odo code we don't do it AFAIK. What's the rationale here?

I have not looked at full code so maybe I'm missing the context here.

Reply:
https://github.com/openshift/odo/pull/4742/files#diff-71f82b8ba7b20d4531fabf0bf878e138fa960894290d65d12e5804c162e39440R55-R61
https://github.com/openshift/odo/pull/4742/files#diff-0fa449830b5dc2b274b7c95a58b5071edc23f1e44af84a99c1fa9e600767829fR85-R88

Reply:
the comment of the function explains this 
```
// To pass operands such as !=, append a ! prefix to the value.// For E.g. map[string]string{"app.kubernetes.io/managed-by": "!odo"}// Using != operators also means that resource will be filtered even if it doesn't have the key.// So a resource not labelled with key "app.kubernetes.io/managed-by" will also be returned.
```

Reply:
>@valaparthvi I'm not sure why someone would try to set "app.kubernetes.io/managed-by": "!odo" as a label to the Kubernetes resources created by odo. At least in the odo code we don't do it AFAIK. What's the rationale here?

I am using it to filter the resources(deployments/dc) not managed by odo. Hence `!odo` part.

Reply:
>e.g. someone did !!odo the logic of = and != is dependent on ! and also it doesn't come as natural to the reader of the code. they have to go to the function and read its comment ( which is fine ) but says that there is room for improvement.
Also the Label approach allows for simpler extending in future.

Okay, that makes sense. But implementing this would probably affect a lot of other tests, I would much rather work on that in a different PR.

Reply:
if you change `strings.Replace(v, "!", "", -1)` to `strings.Replace(v, "!", "", 1)` wouldn't it the problem of doing `!!odo`?

Reply:
@kadel I am not sure I understand your problem correctly.

Reply:
I am good from my end, please address @kadel's question and then remove the hold

Reply:
Just realized something. `!` can't be present in a label or its value.
I think that we are ok with this. front. 

But I think that I'm really missing here is a unit test for this function :-( It became much more complicated than it used to be, and it definetly deserves unit test

Reply:
@kadel Unit tests! That can be easily fixed! I'll get to it.

@girishramnani I looked at how k8s.io/apimachinery does this label selector thing and it is a little complicated, they have special interface and methods for it. See here - https://github.com/openshift/odo/blob/89ac67e425bffe8c400645bca933d687b5ac8b19/vendor/k8s.io/apimachinery/pkg/fields/selector.go#L430

I will rely on Unit tests for now, and hopefully we will not run into corner cases, or catch them if we do.